vortex-gpu/src/rle_decompress.rs

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+use std::sync::Arc;
+
+use cudarc::driver::{CudaContext, DeviceRepr, LaunchConfig, PushKernelArg, ValidAsZeroBits};
+use vortex_array::arrays::PrimitiveArray;
+use vortex_array::validity::Validity;
+use vortex_array::{ArrayRef, IntoArray, ToCanonical};
+use vortex_buffer::BufferMut;
+use vortex_dtype::{NativePType, UnsignedPType, match_each_native_ptype};
+use vortex_error::{VortexExpect, VortexResult, vortex_err, vortex_panic};
+use vortex_fastlanes::RLEArray;
+
+use crate::take::cuda_take_kernel;
+
+pub fn cuda_rle_decompress(
+    rle: &RLEArray,
+    ctx: Arc<CudaContext>,
+) -> VortexResult<Option<ArrayRef>> {
+    match_each_native_ptype!(rle.values().dtype().as_ptype(), |V| {
+        // RLE indices are always u16 (or u8 if downcasted).
+        match rle.indices().dtype().as_ptype() {
+            PType::U8 => cuda_rle_decompress_typed(
+                rle.indices().to_primitive().as_slice::<u8>(),
+                rle.values().to_primitive().as_slice::<V>(),
+                ctx,
+            )
+            .map(Some),
+            PType::U16 => cuda_rle_decompress_typed(
+                rle.indices().to_primitive().as_slice::<u16>(),
+                rle.values().to_primitive().as_slice::<V>(),
+                ctx,
+            )
+            .map(Some),
+            _ => vortex_panic!(
+                "Unsupported index type for RLE decoding: {}",
+                rle.indices().dtype().as_ptype()
+            ),
+        }
+    })
+}
+
+fn cuda_rle_decompress_typed<Values, Indices>(
+    indices: &[Indices],
+    values: &[Values],
+    ctx: Arc<CudaContext>,
+) -> VortexResult<ArrayRef>
+where
+    Values: NativePType + DeviceRepr + ValidAsZeroBits,
+    Indices: UnsignedPType + DeviceRepr,
+{
+    assert_eq!(indices.len() % 1024, 0);
+
+    let kernel_func = cuda_take_kernel::<Indices, Values>(false, ctx.clone())?;
+    let num_chunks =
+        u32::try_from(indices.len().div_ceil(1024)).vortex_expect("num chunks overflow");
+    let stream = ctx.default_stream();
+
+    let cu_values = stream
+        .memcpy_stod(values)
+        .map_err(|e| vortex_err!("Failed to copy to device: {e}"))?;
+    let cu_indices = stream
+        .memcpy_stod(indices)
+        .map_err(|e| vortex_err!("Failed to copy to device: {e}"))?;
+
+    let mut cu_out = unsafe {
```

Comment:
In the original RLE impl this returns the number of values in the `values` array, not the padded indices.

Reply:
there will be as many entries as there's indices, then we need to adjust for offset and length but this has to be overallocated to the indices length to avoid out of bounds writes